pilot/pkg/proxy/envoy/v2/ads.go

```diff
@@ -603,6 +618,214 @@ func (s *DiscoveryServer) StreamAggregatedResources(stream ads.AggregatedDiscove
+				continue
+			}
+			// too verbose - sent immediately after EDS response is received
+			adsLog.Debugf("ADS:LDS: REQ %s %v", con.ConID, peerAddr)
+			con.LDSWatch = true
+			err := s.pushLds(con, s.globalPushContext(), true, versionInfo())
+			if err != nil {
+				return err
+			}
+			if pilot.WaitAck {
+				_ = s.waitAck(con)
+			}
+
+		case RouteType:
+			routes := discReq.GetResourceNames()
+			if routes == nil {
+				adsLog.Warnf("ADS:RDS: ACK PROTOCOL ERROR %v %s (%s) %v", peerAddr, con.ConID, con.modelNode.ID, discReq.String())
+				continue
+			}
+			var sortedRoutes []string
+			sort.Strings(routes)
+			sortedRoutes = routes
+			con.Routes = sortedRoutes
+			adsLog.Debugf("ADS:RDS: REQ %s %s  routes: %d", peerAddr, con.ConID, len(con.Routes))
+			err := s.pushRoute(con, s.globalPushContext())
+			if err != nil {
+				return err
+			}
+			if pilot.WaitAck {
+				_ = s.waitAck(con)
+			}
+
+		case EndpointType:
+			clusters := discReq.GetResourceNames()
+			// clusters and con.Clusters are all empty, this is not an ack and will do nothing.
+			if len(clusters) == 0 && len(con.Clusters) == 0 {
+				adsLog.Warnf("ADS:EDS: unexpected %v %s %v", peerAddr, con.ConID, discReq.String())
+				continue
+			}
+			if len(clusters) == len(con.Clusters) {
+				sort.Strings(clusters)
+				sort.Strings(con.Clusters)
+
+				// Already got a list of endpoints to watch and it is the same as the request, this is an ack without nonce ?
+				if reflect.DeepEqual(con.Clusters, clusters) {
+					adsLog.Warnf("ADS:EDS: ACK without nonce %s %s (%s) %s %s", peerAddr, con.ConID, con.modelNode.ID, discReq.VersionInfo, discReq.ResponseNonce)
+					continue
+				}
+			}
+
+			for _, cn := range con.Clusters {
+				s.removeEdsCon(cn, con.ConID, con)
+			}
+
+			for _, cn := range clusters {
+				s.addEdsCon(cn, con.ConID, con)
+			}
+
+			con.Clusters = clusters
+			adsLog.Debugf("ADS:EDS: REQ %s %s clusters: %d", peerAddr, con.ConID, len(con.Clusters))
+			err := s.pushEds(s.globalPushContext(), con, true, nil)
+			if err != nil {
+				return err
+			}
+			if pilot.WaitAck {
+				_ = s.waitAck(con)
+			}
+
+		default:
+			adsLog.Warnf("ADS: Unknown watched resources %s", discReq.String())
+		}
+
+	}
+}
+
+func (s *DiscoveryServer) handleAck(con *XdsConnection, discReq *xdsapi.DiscoveryRequest, peerAddr string) {
+	con.mu.Lock()
+	switch discReq.TypeUrl {
+	case ClusterType:
+		con.ClusterNonceAcked = discReq.ResponseNonce
+	case ListenerType:
+		con.ListenerNonceAcked = discReq.ResponseNonce
+	case RouteType:
+		routes := discReq.GetResourceNames()
+		routeNonceSent := con.RouteNonceSent
+		routeVersionInfoSent := con.RouteVersionInfoSent
+		if routeNonceSent != discReq.ResponseNonce {
+			adsLog.Debugf("ADS:RDS: Expired nonce received %s %s (%s), sent %s, received %s",
+				peerAddr, con.ConID, con.modelNode.ID, routeNonceSent, discReq.ResponseNonce)
+			rdsExpiredNonce.Inc()
+		}
+		if discReq.VersionInfo == routeVersionInfoSent {
```

Comment:
https://github.com/istio/istio/issues/10105 has a lot of explanations - I tried to keep the code the same as in the original, but if we can clean up it would be better.